File: app/schemas/ops.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import Field

from app.schemas.base import CamelModel


class OpsMCPInstallPreviewRequest(CamelModel):
    link: str
    conversation_id: str | None = None


class OpsMCPInstallExecuteRequest(CamelModel):
    session_id: str
    step_id: str


class OpsMCPInstallStep(CamelModel):
    id: str
    name: str
    title: str
    status: str
    requires_confirm: bool = True
    detail: str = ""
    result: dict[str, Any] = Field(default_factory=dict)
    error_category: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None


class OpsMCPParsedConfig(CamelModel):
    source_type: str
    name: str
    description: str
    transport_type: str
    endpoint_or_command: str
    advanced_config: dict[str, Any] = Field(default_factory=dict)
    raw: dict[str, Any] = Field(default_factory=dict)


class OpsMCPEnvProbeItem(CamelModel):
    command: str
    available: bool
    path: str | None = None
    version: str | None = None
    detail: str = ""


class OpsMCPInstallSession(CamelModel):
    id: str
    link: str
    conversation_id: str | None = None
    status: str
    summary: str = ""
    parsed_config: OpsMCPParsedConfig
    env_report: list[OpsMCPEnvProbeItem] = Field(default_factory=list)
    steps: list[OpsMCPInstallStep] = Field(default_factory=list)
    server_id: str | None = None
    created_at: datetime
    updated_at: datetime


class OpsMCPInstallPreviewResponse(CamelModel):
    session: OpsMCPInstallSession


class OpsMCPInstallExecuteResponse(CamelModel):
    session: OpsMCPInstallSession
    step: OpsMCPInstallStep

File: src/types/ops.ts
export type OpsMCPInstallStepStatus = 'pending' | 'running' | 'passed' | 'failed'

export interface OpsMCPInstallStep {
  id: string
  name: string
  title: string
  status: OpsMCPInstallStepStatus | string
  requiresConfirm: boolean
  detail: string
  result?: Record<string, unknown>
  errorCategory?: string
  startedAt?: string
  finishedAt?: string
}

export interface OpsMCPParsedConfig {
  sourceType: string
  name: string
  description: string
  transportType: 'http' | 'stdio' | string
  endpointOrCommand: string
  advancedConfig?: Record<string, unknown>
  raw?: Record<string, unknown>
}

export interface OpsMCPEnvProbeItem {
  command: string
  available: boolean
  path?: string
  version?: string
  detail: string
}

export interface OpsMCPInstallSession {
  id: string
  link: string
  conversationId?: string
  status: string
  summary: string
  parsedConfig: OpsMCPParsedConfig
  envReport: OpsMCPEnvProbeItem[]
  steps: OpsMCPInstallStep[]
  serverId?: string
  createdAt: string
  updatedAt: string
}

export interface OpsMCPInstallPreviewRequest {
  link: string
  conversationId?: string
}

export interface OpsMCPInstallExecuteRequest {
  sessionId: string
  stepId: string
}

export interface OpsMCPInstallPreviewResponse {
  session: OpsMCPInstallSession
}

export interface OpsMCPInstallExecuteResponse {
  session: OpsMCPInstallSession
  step: OpsMCPInstallStep
}

export interface OpsCommandPreview {
  command: string
  argv: string[]
  cwd: string
  riskLevel: string
  requiresConfirm: boolean
  notes: string[]
}

export interface OpsCommandResult {
  exitCode: number
  stdout: string
  stderr: string
  durationMs: number
}

export interface OpsCommandSession {
  id: string
  conversationId?: string
  status: string
  summary: string
  preview: OpsCommandPreview
  result?: OpsCommandResult | null
  createdAt: string
  updatedAt: string
  startedAt?: string | null
  finishedAt?: string | null
  errorCategory?: string | null
  errorMessage?: string | null
  metadata?: Record<string, unknown>
}

export interface OpsCommandPreviewRequest {
  command: string
  cwd?: string
  conversationId?: string
}

export interface OpsCommandExecuteRequest {
  sessionId: string
}

export interface OpsCommandPreviewResponse {
  session: OpsCommandSession
}

export interface OpsCommandExecuteResponse {
  session: OpsCommandSession
}
